src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -147,6 +152,37 @@ def warn_python_repos(option: str) -> None:
     return MaybeWarnPythonRepos()
 
 
+@rule_helper
+async def _setup_pip_args_and_constraints_file(
+    python_setup: PythonSetup, *, resolve_name: str
+) -> tuple[list[str], Digest, FileDigest | None]:
+    extra_args = []
+    extra_digests = []
+    constraints_file_digest: None | FileDigest = None
+
+    if python_setup.no_binary or python_setup.only_binary:
+        pip_args_file = "__pip_args.txt"
+        extra_args.extend(["-r", pip_args_file])
+        pip_args_file_content = "\n".join(
+            [f"--no-binary {pkg}" for pkg in python_setup.no_binary]
```

Comment:
This was pre-existing, but changing the order of the `python_setup` args produces a new file here and the subsequent downstream invalidations that flow when it need not, `--no-binary` / `--only-binary` is not an order-sensitive thing.

Reply:
> This was pre-existing

My next PR will likely be reworking this mechanism to be resolve-specific, so I can clean that up there.